### PR TITLE
Fix tem object out of scope issue

### LINKED ISSUE #8674: tiflash crash with 'Received signal Bus error' during query
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```sql
 create table t01 ( `COL1` enum('^YSQT0]V@9TFN>^WB6G?NG@S8>VYOM;BSC@<BCQ6'),  `COL2` mediumint(41) DEFAULT NULL,    `COL3` year(4) DEFAULT NULL,  KEY `U_M_COL4` (`COL1`,`COL2`), KEY `U_M_COL5` (`COL3`,`COL2`)  );
alter table t01 set tiflash replica 1;
insert into t01 values ('^YSQT0]V@9TFN>^WB6G?NG@S8>VYOM;BSC@<BCQ6', -1881752, 1986);
 SELECT * FROM t01 where col2 = -1881752 and col2 * -1881752 != 8366212;
```

### 2. What did you expect to see? (Required)
query success
### 3. What did you see instead (Required)
```log
query_id:4, server_id:289, start_ts:446876954885881857,task_id:1>"] [thread_id=66]
[2024/01/08 15:51:37.895 +08:00] [INFO] [MPPTask.cpp:542] ["task starts running, time cost in schedule: 0 ms, time cost in preprocess: 2 ms"] [source="MPP<gather_id:1, query_ts:1704700297887677000, local_query_id:4, server_id:289, start_ts:446876954885881857,task_id:1>"] [thread_id=66]
[2024/01/08 15:51:37.895 +08:00] [INFO] [SegmentReadTaskScheduler.cpp:66] ["Added, pool_id=3 block_slots=8 segment_count=1 pool_count=1 cost=1.000us do_add_cost=1.000us"] [source="MPP<gather_id:1, query_ts:1704700297887677000, local_query_id:4, server_id:289, start_ts:446876954885881857,task_id:1> table_id=109"] [thread_id=66]
[2024/01/08 15:51:37.898 +08:00] [ERROR] [BaseDaemon.cpp:367] [########################################] [source=BaseDaemon] [thread_id=67]
[2024/01/08 15:51:37.898 +08:00] [ERROR] [BaseDaemon.cpp:368] ["(from thread 34) Received signal Bus error: 10(10)."] [source=BaseDaemon] [thread_id=67]
[2024/01/08 15:51:37.898 +08:00] [ERROR] [BaseDaemon.cpp:427] ["Invalid address alignment."] [source=BaseDaemon] [thread_id=67]
[2024/01/08 15:51:37.901 +08:00] [ERROR] [BaseDaemon.cpp:560] ["\n     0x1a562f18c\t___simple_bprintf [libsystem_platform.dylib+6445871500]"] [source=BaseDaemon] [thread_id=67]
```
```sql
[15:51:37]TiDB root:test>  SELECT * FROM t01 where col2 = -1881752 and col2 * -1881752 != 8366212;
(1105, 'rpc error: code = Unavailable desc = error reading from server: EOF')
```
### 4. What is your TiFlash version? (Required)
```sql
[15:56:45]TiDB root:test> select type,version,git_hash  from information_schema.cluster_info;
+---------+-------------+------------------------------------------+
| type    | version     | git_hash                                 |
+---------+-------------+------------------------------------------+
| tidb    | 7.6.0-alpha | 9b0fd9ea299266da70456f6e6077ed14bd191cfc |
| pd      | 7.6.0-alpha | c3ad361486ef29fb1a9001f5374788ad741d5616 |
| tikv    | 7.6.0-alpha | e0d70726b332a33e503c3f2addc66b9794303aea |
| tiflash | 7.6.0-alpha | 383e1bd9d3a793f6adb84156639febc68739ea63 |
+---------+-------------+------------------------------------------+
```

### PR DESCRIPTION
**Pull request by yibin87**
### What problem does this PR solve?

Issue Number: close #8674 

Problem Summary:
When encode enum column, we get the enum name from DataTypeEnum's getNameForValue. However, the return value is a temporary object, and its life scope is limited by the construction of TiDBEnum:
https://github.com/pingcap/tiflash/blob/b9468ccfaa0da1e004b36acfb081b3da4ff2d07e/dbms/src/Flash/Coprocessor/ArrowColCodec.cpp#L350C1-L352C99,
And the name field in TiDBEnum is a constant reference also, which points to the temporary object. 
Confirmed this root cause by two steps:
1. When the crash happen, the getNameForValue still can return the correct name, that means the DataTypeEnum instance memory is not released
2. The illegal StringRef's char pointer is different from the one inside getNameForValue.

Still not sure why v7.5.0 is not affected by this issue, debug version seems not affected by the issue either.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Flash/Coprocessor/ArrowColCodec.cpp [hunk 1]
```diff
@@ -347,9 +347,8 @@ void flashEnumColToArrowCol(
                 enum_value,
                 enum_value_size);
 
-        TiDBEnum ti_enum(
-            enum_value,
-            enum_type->getNameForValue(static_cast<const DataTypeEnum16::FieldType>(enum_value)));
+        const auto & enum_name = enum_type->getNameForValue(static_cast<const DataTypeEnum16::FieldType>(enum_value));
+        TiDBEnum ti_enum(enum_value, enum_name);
```
Comment (xzhangxian1008) [inline dbms/src/Flash/Coprocessor/ArrowColCodec.cpp:351]: I think it's better to change the `TiDBEnum`'s `name` type to `StringRef`, as the copy of `StringRef` is very small.

**Review by windtalker — APPROVED** [review]
LGTM

File: tests/fullstack-test/issues/issue_8767.test [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+func> wait_table test t1
+
+mysql> set session tidb_isolation_read_engines='tiflash'; set tidb_enforce_mpp=1; SELECT * FROM test.t1 where col2 = -1881752 and col2 * -1881752 != 8366212;
++------------------------------------------+----------+------+
+| COL1                                     | COL2     | COL3 |
++------------------------------------------+----------+------+
+| ^YSQT0]V@9TFN>^WB6G?NG@S8>VYOM;BSC@<BCQ6 | -1881752 | 1986 |
++------------------------------------------+----------+------+
+
+mysql> drop table if exists test.t1
```
[the remark this answers is not in the document]
Comment (yibin87) [reply tests/fullstack-test/issues/issue_8767.test:28]: Added

File: dbms/src/Flash/Coprocessor/ArrowColCodec.cpp (hunk 1 above)
Comment (yibin87) [reply dbms/src/Flash/Coprocessor/ArrowColCodec.cpp:351]: Sounds good, change it.

**Review by xzhangxian1008 — APPROVED** [review]
LGTM

**Pull request merged**